src/app/zap-templates/templates/chip/im-encoder-h.zapt

```diff
@@ -0,0 +1,46 @@
+
```

Comment:
Can we reuse `src/app/zap-templates/templates/chip/CHIPClusters.zapt` instead of adding a new interface to use the IM code ? 

Basically I would like to get rid of `src/app/encoder.cpp` at some point, but in the current state some apps and `chip-tool` still use it so I have kept it instead of directly generating the right code into the `CHIPClusters-src.zapt` template.

There are a few differences between the 2 templates but it can probably be unified.

The current namespace defined by `CHIPClusters.h` is `chip::Controller::{{asCamelCased name false}Cluster` while this one is under `chip::app::cluster::{{asCamelCased name false}}`. It makes more sense to use `chip::app::cluster` imho. FWIW I have reuse the namespace `chip::Controller` since the initial landing of `CHIPCluster` was using it and I was trying to minimise the changes.

`CHIPCluster` makes the cluster classes to inherit from `src/controller/CHIPCluster.h` which has the `endpointId` and the `clusterId` has protected member.  This template also uses `groupId`. It can be added to `ClusterBase`.

`CHIPCluster` uses `Callback::Callback<> * onCompletion` as the first command parameter. A new method can be generated that takes the same arguments list but replace the callback by `chip::app::Command * ZCLcommand`

As an example for the `GoToPercent`method of `BarrierControl`, 2 methods can be generated until everything is switched to the IM version:
 ```
CHIP_ERROR BarrierControlCluster::BarrierControlGoToPercent(Callback::Callback<> * onCompletion, uint8_t percentOpen);
CHIP_ERROR BarrierControlCluster::BarrierControlGoToPercent(chip::app::Command * ZCLcommand, uint8_t percentOpen);
```

`CHIPCluster` does not uses `const` for the parameters while this template does. Sounds good to me to use it. 

About `InitCluster` and `ShutdownCluster`, the way their work is still unclear to me.
My current understanding of the code is that `InitCluster` is used to register callbacks methods when a method is received ? If so it means that it bypass all the ZCL imported code from `src/app` which looks wrong to me. But maybe I have misunderstood the code.

I would like to discuss those points before going into a more granular review.

Reply:
Took a glance of the CHIPCluster.zapt here are some ideas:
1. I guess the encoders will continue exist, since the IM allows sending multiple commands in one packet, so we have to Encode the command and then Send it. However, the controller can provide a convenient api for encode and send command. 
2. Do you mean that add `EncodeXXXCommand` to `XXXCluster` class?
3. `InitCluster` does registerss callbacks and current ZCL implementation does not works since they does not go through the MessageLayer, which is required by the InvokingCommands, since the imported ZCL does not support MessageLayer api, it does not work currently. So I adds a build flag so current code still works. This is a follow up and we can deal with it later.

Reply:
> Took a glance of the CHIPCluster.zapt here are some ideas:
> 
> 1. I guess the encoders will continue exist, since the IM allows sending multiple commands in one packet, so we have to Encode the command and then Send it. However, the controller can provide a convenient api for encode and send command.
> 2. Do you mean that add `EncodeXXXCommand` to `XXXCluster` class?

Yes. 

I think it would be better if we all work on the same class instead of having some code that uses `chip::Controller::XXXCluster` class and some other that uses `chip::app:cluster::XXX` class.

If we do that, my understanding is that the core difference between the current `XXXCluster` methods and what you did as specified by the IM, is that in the first case the command is sent directly along with a callback per command, while in your case a `chip::app:Command` parameter will be passed along to multiple clusters before a higher level APIs decides to send it.


> 3. `InitCluster` does registerss callbacks and current ZCL implementation does not works since they does not go through the MessageLayer, which is required by the InvokingCommands, since the imported ZCL does not support MessageLayer api, it does not work currently. So I adds a build flag so current code still works. This is a follow up and we can deal with it later.

I guess my main concern is that this patch moves handling of packets received from `src/app` to the application side directly, effectively bypassing the ZCL internals. And the current patch does not makes it clear to me what is the plan for handling received packets, dispatching them to the ZCL internals, before dispatching them to the possible application handlers.

Can we move this encoding/decoding scheme forward without the `ExchangeMgr` at the moment ?

For example can we send the IM encoded packets over the wire using our network code, decode the IM packet inside https://github.com/project-chip/connectedhomeip/blob/d05149c42866c6e301f1f976301b2c774be0d443/src/app/server/DataModelHandler.cpp#L46 manually building APS frame from the content of the message ?
That would be a first step towards replacing the current encoding code. 
From here (maybe in a different PR) one may be able to pass the retrieved `chip::app:Command` packet to `emberAfProcessMessage` instead of the raw `uint8_t *` parameter.

What do you think ?

Reply:
In this PR, the Handlers for invoking commands is missing, and they should placed in `src/app/clusters`, the handlers in `examples/lighting/linux` is a temporary solution so the linux example app can use the IM for ZCL.

Some follow up PRs are needed for the actual command handlers.